File: solutions/ruby/problem-00021.rb
#!/usr/bin/env ruby
# [minhuang@mocha:ruby]$ time ruby problem-00021.rb 
# 31626
#
# real  0m1.385s
# user  0m1.356s
# sys 0m0.004s
def divisors(n)
  d = (2..Math.sqrt(n)).select { |q| n%q == 0 }
  d = d + d.map { |q| n / q } + [1]
  d.uniq
end

def sigma(n)
  divisors(n).inject(:+)
end

limit = 10000
result = (2..limit).map { |n| [n, sigma(n)] }
result = result.select { |t| t.first > t.last && t.first == sigma(t.last) }
result = result.map { |t| t.first + t.last }
result = result.inject(:+)
puts result

File: haskell/problem-00021.hs
{-
[minhuang@mocha:haskell]$ time runghc problem-00021.hs 
31626

real  0m2.343s
user  0m2.216s
sys 0m0.056s
-}
import Data.List (nub)

divisors n = let sqn = floor . sqrt . fromIntegral $ n
                 a = filter (\q -> mod n q == 0) [2..sqn]
                 b = map (\q -> div n q) a
             in nub ([1] ++ a ++ b)

sigma n = sum . divisors $ n

range = [2..10000]
result = let tuples = zip range (map sigma range)
             amicables = filter (\(a,b) -> a > b && a == sigma b) tuples
         in sum $ map (\(a,b) -> a+b) amicables

main = print result         
